### PR TITLE
Hiera5

### PR DESCRIPTION
**Pull request by naeem98**
- Added epp for hiera 5 
- Added hiera_version, hiera5_defaults and hierarchy (Array[Hash]) for hiera5
- hiera_version defaults to version 3
- Added custom types for hiera5_defaults and hierarchy for hiera 5

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -62,6 +62,20 @@ class { 'hiera':
...
+          hiera_version   =>  '5',
+          hiera5_defaults =>  {"datadir" => "data", "data_hash" => "yaml_data"},
+          hierarchy       =>  [
+                                {"name" =>  "Virtual yaml", "path"  =>  "virtual/%{::virtual}.yaml"},
+                                {"name" =>  "Nodes yaml", "paths" =>  ['nodes/%{::trusted.certname}.yaml', 'nodes/%{::osfamily}.yaml']},
+                                {"name" =>  "Default yaml file", "path" =>  "common.yaml"},
+                              ],
+}
+```
+# Note: For Hiera version 5 when calling the class, please remember to pass '5' to 'hiera_version' as in the example above.
```
Comment (bastelfreak) [inline README.md]: this is a bit huge. Maybe we should only make it bold?

File: manifests/init.pp [hunk 2]
```diff
@@ -40,52 +55,53 @@
...
-  $eyaml_gpg_source        = undef,
-  $eyaml_gpg               = false,
-  $eyaml_gpg_recipients    = undef,
-  $eyaml_pkcs7_private_key = undef,
-  $eyaml_pkcs7_public_key  = undef,
-
-  $ruby_gpg_name           = 'ruby_gpg',
-  $ruby_gpg_version        = undef,
-  $ruby_gpg_source         = undef,
+  Variant[Array, Array[Hash]] $hierarchy    = $::hiera::params::hierarchy,
```
Comment (bastelfreak) [inline manifests/init.pp]: Datatypes \o/

File: manifests/init.pp [hunk 3]
```diff
@@ -214,9 +230,31 @@
...
   # - $extra_config
+
+  # Hiera 5 additional parameters:
+  # - hiera_version (String)
+  # - hiera5_defaults (Hash)
+  # - hierarchy (Array[Hash])
+
+  # Determine hiera version
+  case $hiera_version {
+    '5':          { if $hierarchy !~ Hiera::Hiera5_hierarchy {
```
Comment (bastelfreak) [inline manifests/init.pp]: there is a but much whitespace beforethe initla {, can you reformat this?

File: templates/hiera.yaml.epp [hunk 4]
```diff
@@ -0,0 +1,73 @@
+<%- |
```
Comment (bastelfreak) [inline templates/hiera.yaml.epp]: this is a huge template. Can you add some sort of tests for it?

**Comment by bastelfreak** [comment]
Hi @naeem98, thanks for the great PR!

Can you please take a look at commit 650f144? It uses an email address that isn't associated with your github account.

File: README.md (hunk 1 above)
Comment (naeem98) [inline README.md]: Changed to bold

File: manifests/init.pp (hunk 2 above)
Comment (naeem98) [inline manifests/init.pp]: :)

File: manifests/init.pp (hunk 3 above)
Comment (naeem98) [inline manifests/init.pp]: Fixed extra white spaces

File: templates/hiera.yaml.epp (hunk 4 above)
Comment (naeem98) [inline templates/hiera.yaml.epp]: Wrote some tests. Please take a look and let me know if you think that we should more tests.

File: spec/classes/hiera_spec.rb [hunk 5]
```diff
@@ -528,5 +528,117 @@
         it { is_expected.to contain_class('hiera::deep_merge') }
       end
     end
+    context 'hiera version 5' do
+      let (:facts) do
+        {
+          puppetversion: Puppet.version,
```
Comment (bastelfreak) [inline spec/classes/hiera_spec.rb]: can you migrate this to rspec-puppet-facts? An example is available at: https://github.com/voxpupuli/puppet-zabbix/blob/master/spec/classes/database_mysql_spec.rb#L8-L12

Feel free to join our IRC channel #voxpupuli on freenode if you've any questions.

**Comment by bastelfreak** [comment]
You can fix probably most of the rubocop errors with:

```bash
bundle exec rubocop -c .rubocop.yml -D -S -a
```

File: spec/classes/hiera_spec.rb (hunk 5 above)
Comment (naeem98) [inline spec/classes/hiera_spec.rb]: Updated with rspec-puppet-facts. It looks good now.

**Pull request merged**